### ISSUE TITLE
gm-control should be behind a sidecar

### ISSUE THREAD

**Issue by justincely** [DecipherNow/helm-charts#749]
now that HTTP2 support is full in the mesh (no downgrade) we can put gm-control behind a managed sidecar so that it registers in the dashboard, collects metrics, and can be accessed outside via edge

**Comment by chrisbsmith** [comment]
@justincely I assume that we still have a chicken-egg problem here and that we still need a control svc so control-api can find it?  Otherwise, how do we configure the control sidecar dynamically?

**Comment by rahug20** [comment]
Based on talks with him, we might have to configure static sidecar configuration if dynamic discovery does not work

**Comment by justincely** [comment]
I don't believe this is chicken and egg actually.  I believe there is only a single connection that is stuck in that paradigm and that would be putting a sidecar _between_ `gm-control` and `gm-control-api`.  In that gm-control needs to connect to `gm-control-api` in order to get the valid configurations for the sidecar that would accept that same incoming connection.  Not gonna work well.

Talking about the issue at hand though, i think it works fine.

1. deploy `gm-control-api`
1. deploy `gm-proxy` + `gm-control`
1. `gm-proxy` connects to `gm-control` and waits for it's configuration
2. `gm-control` gets it's own sidecar node from service discovery (k8s, consul, aws APIs)
1. `gm-control` connects directly to `gm-control-api`, gets configs for it's attached `gm-proxy`
1. `gm-control` feeds configurations to the attached `gm-proxy`, it gets setup to receive connections
1. sidecars in the mesh connect to the `gm-proxy` in front of `gm-control` to discover their configs

**Comment by chrisbsmith** [comment]
ok, so this is going to be a pretty big change @rahug20 .  We probably need to add a new route to all services to find control at `https://localhost:10909/control` and remove the control service.  This also changes our paradigm on how to add services to the mesh. Currently, we configure the `XDS_SERVER` to `control.greymatter.svc:5555`.  We will need to update our service templates to add the new route and the new values in the configs.  

We actually probably don't want to deprecate the control svc quite yet until we have a strong handle on this.

**Comment by justincely** [comment]
@chrisbsmith actually it's a little simpler than that because you can't remove the k8s service.  I missed this detail in the above, but have added it now.  

Basically you do still need the k8s service or other load balancer in front of control.  Just now it points to the proxy rather than control directly.  Sidecars can't get routes+cluster to point to gm-control without first connecting to gm-control, so _that_ part is still chicken/egg.  This change doesn't solve that, but rather allows metrics to be recorded for control as well as RBAC, observables, etc

**Comment by rahug20** [comment]
@chrisbsmith You me

> ok, so this is going to be a pretty big change @rahug20 . We probably need to add a new route to all services to find control at `https://localhost:10909/control` and remove the control service. This also changes our paradigm on how to add services to the mesh. Currently, we configure the `XDS_SERVER` to `control.greymatter.svc:5555`. We will need to update our service templates to add the new route and the new values in the configs.
> 
> We actually probably don't want to deprecate the control svc quite yet until we have a strong handle on this.

You mean find control on port `8080` via the sidevar?

**Comment by rahug20** [comment]
PR: https://github.com/greymatter-io/helm-charts/pull/781/files

**Comment by rahug20** [comment]
Latest PR: https://github.com/greymatter-io/helm-charts/pull/785

**Comment by moosecanswim** [comment]
@justincely @chrisbsmith, I just wanted to point out that even if we put control and control api behind sidecars we should preserve an svc for control-api so in the case that the mesh is in a bad state (causing edge to be compromised) we can still access the api using a port-forward

**Comment by chrisbsmith** [comment]
@moosecanswim note that you don't need an svc to use port-forward.  you can port forward directly to the pod.  
```console
kubectl port-forward -n greymatter control-api-0 5555
```

**Comment by justincely** [comment]
@moosecanswim I believe the PR does still retain the SVC.  Though the proxy can be used in front of control-api for most service and user connections, it cannot be used in the connection between `gm-control` and `gm-control-api`.  For that connection, a svc is still required
